### ISSUE TITLE
md-dialog: close dialog on location change

### ISSUE THREAD

**Issue by cnpal** [angular/material#3418]
You should consider closing dialogs when the location changes (e.g. $location.path(...) or browser back-button). Otherwise weird stuff starts to happen if you don't do that as the underlying controller is gone.

I fixed that in my base controller using: 
`$rootScope.$on('$routeChangeStart', function() { $mdDialog.cancel(); })`

A possible fix has already been implemented for angular-ui, so you might want to apply that for angular material:

https://github.com/angular-ui/bootstrap/commit/474ce52e8eadee763987ebd930cf790d4bcd6d97

**Comment by bradrich** [comment]
@ThomasBurleson What do you think about implementing this into the logic for MdPanels instead of adding more logic into MdDialog? @ErinCoughlan and I are working on functionality for multiple MdPanels that would allow for methods to be run on more than one MdPanel that would solve this issue.

**Comment by ErinCoughlan** [comment]
@bradrich If it were added to MdPanel, it would be another configuration option. I'm not against it, but it can't happen with every panel as opposed to dialogs where it's possible to do always.

@ThomasBurleson Something to consider with this FR is to also add a callback onClose so that when the user tries to navigate back and there are unsaved edits, the user can be alerted and the close/navigation event can be cancelled. This is definitely needed for MdPanel.

**Comment by ivanzvakoms** [comment]
I agree with @bradrich that we need to move all possible components to mdPanel and implement this fix there only, because the same issue exists in mdPanel for now.

**Comment by alirezamirian** [comment]
> $rootScope.$on('$routeChangeStart', function() { $mdDialog.cancel(); })
This will cause weird bugs if you have **dialog states**, states which open a dialog on their entrance event.

**Comment by dbrgn** [comment]
ngDialog has implemented something like this: https://github.com/likeastore/ngDialog#closebynavigation-boolean
